### DISCUSSION TITLE
Timeout_ops & timeout_transport for long commands

### DISCUSSION THREAD

**Discussion by lehoangkhanh2008** [carlmontanari/scrapli#118] [Q&A] [answered]
Hello  Carl & everyone,

Thanks for your great effort to bring this library.

I'm a network administrator & doing some research on Scrapli. I was familiar with Netmiko before.

One use case that I'd usually face is that pushing long batch of configuration(may up to 100 commands or even greater) or working on slow connection. In such case, I measure in SSH terminal (like Putty or other tool), it can take upto 3~5 mins to complete all command.

In Netmiko, I adjust delay factor & max loop to deal with this situation.

I know in Scrapli, we can do the same with Timeout_ops & timeout_transport, however I'm not really sure about its behavior.
By doing some debug, I found that default value Timeout_ops & timeout_transport  is 30s.

My question is that 30s timeout is measured since Scrapli send first command OR from the last time that Scrapli receive any output from device?

What is the best value of Timeout_ops & timeout_transport  in this case ?

Also, it would be great if you can tell me where in you source code that you implement the timeout mechnism? I tried to find but don't see it.

Best Regards,
Khanh

**Comment by carlmontanari** [comment, marked as answer]
Hey Khanh!

You are correct -- the default value for both `timeout_ops` and `timeout_transport` is 30s. There is also a `timeout_socket` with a default value of 15s. 

The `timeout_socket` value governs the "initial" connection basically -- or opening the socket to the device. The exact way this is handled is dependent upon which transport you use. This argument only matters at the initial spin up of the connection and is ignored beyond that.

Next the `timeout_transport` governs the timeout for each individual "read" operation in the transport (transport meaning, paramiko, ssh2-python, the custom "system" transport that is default, etc.).

Lastly `timout_ops` is basically what it sounds like -- a timeout *per operation*. 

I'll ignore the `timeout_socket` for now since you didn't ask about it directly and its the most variable just because of each of the different types of transports having a slightly different way to use it.

You can see [here](https://github.com/carlmontanari/scrapli/blob/3f8724f877dedc65ade32b3f7aec5d8de1556bc5/scrapli/transport/plugins/system/transport.py#L121-L135) where we decorate the read method of the system transport. So this `timeout_transport` is literally the timeout for every single read operation.

[Here](https://github.com/carlmontanari/scrapli/blob/3f8724f877dedc65ade32b3f7aec5d8de1556bc5/scrapli/channel/sync_channel.py#L398-L444) you can see an example of where we decorate a "channel operation" with the channel timeout decorator. So this is basically the timeout for an *operation* -- an *operation* may end up having lots of reads of the transport. For example you may send "show tech" or something -- obviously the device will take a while to spit out all the data -- the `send_input` method (which is called for you when you use `send_command` at the generic/network driver layer) will keep calling the read method of the transport until it finds the prompt. So you may need to turn up the `timeout_ops` to something (let's jus say 180s) in order to not hit timeouts when doing a "show tech" or other long command.

*Usually* you don't need to turn up the `timeout_transport` as the only time you would hit this timeout is when the device is literally not sending *any* data for >30s (which is not super typical... there's usually at least *something* on the wire).

Hopefully that addresses your main question. As for the "best value" for timeouts... obviously that is dependent on lots of things... that said I have been basically leaving the `timeout_socket` along, and setting `timeout_transport` to `0` lately. By setting the `timeout_transport` to `0` you disable that timeout decorator so everything ends up being handled just by `timeout_ops` which is a lot simpler to tune as you can just decide what value and off ya go! Historically the "in channel" authentication for transports like `telnet` and `system` happened *in* the transport itself -- in this case the `timeout_transport` value was actually much more important as it governed the authentication process... this is no longer the case (as of the 2021.01.30 release). 

So TLDR: probably safe to leave `timeout_socket` alone, set `timeout_transport` to `0`, then just set `timeout_ops` to however long it takes for your longest output to print out on the terminal. 

Oh, almost forgot -- timeout decorators live [here](https://github.com/carlmontanari/scrapli/blob/master/scrapli/decorators.py#L21)... the method of *how* the timeout happens varies based on some things.... asyncio just uses native asyncio `wait_for`, then for others we *prefer* to use a `signal` based timeout as it is super efficient, but we cant do that for `system` transport, so as a last result we have a thread based wrapper to handle timeouts.

HTH
Carl

**Reply by lehoangkhanh2008** [reply]
Perfect! Supper thanks Carl very detail answer. 
Now I'm more confident while using Scrapli.

Good day to you!

Regards,
Khanh
